### PR TITLE
text_input_cursor_rect_for_byte_offset

### PR DESCRIPTION
**Pull request by lukas-jung**
this is the same as #1104 but cleaned up.

### CODE REVIEW

File: internal/core/items/text.rs [hunk 1]
```diff
@@ -571,18 +571,22 @@ impl TextInput {
...
                 reset_preferred_x_pos = false;
 
-                let mut cursor_xy_pos =
-                    window.text_input_position_for_byte_offset(self, last_cursor_pos);
-                cursor_xy_pos.y -= font_height * 0.5;
+                let cursor_rect =
+                    window.text_input_cursor_rect_for_byte_offset(self, last_cursor_pos);
+                let mut cursor_xy_pos = cursor_rect.center();
+
+                cursor_xy_pos.y -= font_height;
```
Comment (ogoffart) [inline internal/core/items/text.rs:589]: Here you also need to special case for the first line. If it becomes negative, we don't want to move the cursor.
Also on the last line we also don't want to move the cursor when going down

File: internal/core/items/text.rs [hunk 2]
```diff
@@ -612,7 +616,7 @@ impl TextInput {
 
         if reset_preferred_x_pos {
             self.preferred_x_pos
-                .set(window.text_input_position_for_byte_offset(self, new_cursor_pos).x);
+                .set(window.text_input_cursor_rect_for_byte_offset(self, new_cursor_pos).origin.x);
```
Comment (ogoffart) [inline internal/core/items/text.rs:619]: I think this should be done in set_cursor_position. It can take ab argument If it should be updated or not

**Review by ogoffart — APPROVED** [review]
Looks good. I just left a few comments that can be done in a follow up

File: internal/core/items/text.rs (hunk 2 above)
Comment (lukas-jung) [reply internal/core/items/text.rs]: that was easy to do, so i added it to this pr.

File: internal/core/items/text.rs [hunk 3]
```diff
@@ -367,7 +367,7 @@ impl Item for TextInput {
                 if self.as_ref().pressed.get() {
                     let clicked_offset =
                         window.text_input_byte_offset_for_position(self, pos) as i32;
-                    self.set_cursor_position(clicked_offset, window);
+                    self.set_cursor_position(clicked_offset, true, window);
 
                     self.preferred_x_pos.set(pos.x);
```
Comment (ogoffart) [inline internal/core/items/text.rs:372]: this is no longer necessary.  
same thing a few lines before.

**Review by ogoffart — APPROVED** [review]

**Pull request merged**